x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -353,6 +366,7 @@ private void createApiKeyAndIndexIt(
                     created,
                     expiration,
                     request.getRoleDescriptors(),
+                    request.getType(),
```

Comment:
Since API key type and request class are currently 1:1, the `getType()` call here (and a couple other places) can be replaced by something like 
```java
request instanceof CreateCrossClusterApiKeyRequest ? ApiKey.Type.CROSS_CLUSTER : ApiKey.Type.REST
```
This means we can possibly drop `type` from the request classes altogether since the class type itself already indicates it. Without the `type`, the two request classes have essentially no difference other than the class type itself. This makes me ponder again on subclassing v.s. just adding an extra `type` field into the existing `CreateApiKeyRequest` because the class hierarchy seems to be rather superficial. For updating API keys, if we also choose to subclass, it probably means two more subclasses, one for single update and one for bulk update (even though cross-cluster does not support bulk update with API, the code always uses bulk updating behind the scene). Given these subclasses have little difference, is the cost benefit here worth it? Maybe having the class hierarchy is more flexible and expandable in future? Or maybe you have a different subclass implementation in mind? Please let me know. Thanks!

Reply:
For sure agreed that the delta between the classes is small and that the hierarchy forces some overhead. We still have some difference in behavior: the validation is different, and I'm also suggesting a difference in serialization behavior in another comment. 

My other motivation for suggesting inheritance is that it creates a stronger conceptual separation between the two. My mental model for these is that they are two conceptually separate things as opposed to one thing that's differently parametrized. 

That said, I took a look at the update requests and I think it could get a little convoluted to have a class hierarchy there. We already have an abstract base class. We could make another split and introduce a AbstractBulkUpdate class with two sub-classes. The REST layer parser could then do the work on converting a conceptually single-target request into a bulk one. That feels like fairly heavy machinery for what we're trying to accomplish though.

I don't have a strong preference here one way or another. I like the conceptual separation we get with the hierarchy but agree that it feels a little heavy-handed, esp. considering the update routes.

If we go with the hierarchy approach, I agree that it makes sense to get rid of the `getType` method, as it's a source of redundancy 👍

Reply:
Thanks! I'll stick to the subclass approach. There is some more work to achieve this for updating. But it is *not* unmanageable.

> If we go with the hierarchy approach, I agree that it makes sense to get rid of the getType method, as it's a source of redundancy

For now, I kept the `getType` method because we otherwise need a translation logic outside of the Request classes. For now it can be a ternary statement. In future, it might need to be a lookup map. Having it as part of the request still helps.